visualize-admin/visualization-tool
# Authentication scaffold

Preliminary work for authentication.

- Ability to login through a Keycloak instance
- Simple login menu (behind a flag)
- Simple profile page listing user charts

## How to test

Right now, this can only be tested locally since we need a Keycloak
instance and we do not have yet the access to the one that cube creator
uses.

1. Run `yarn install` and `yarn db:migrate:dev` to migrate the database
2. Setup the Keycloak instance and configure the application through the env file
   (more information in the README.md).
3. When Keycloak is configured (and a `flag__login=true` is added to the URL), the login menu will display in the header, and
you'll be able to sign in.
4. When you are signed in, when you create a chart, it will be linked to the
currently logged in user.
5. By clicking your name in the login menu, you can go to the scaffold profile page
and see the charts you created.

## Technical information

### Libraries

- Added next-auth for authentication through the provided Keycloak provider
- Added Prisma for easier access to database models
- Added superjson for serialization of "complex" objects through getServerSideProps
  since Prisma parses dates.

### Models

- Config have a user_id column now
- Added User model (not used for authentication, simply to
  link config to users)

### Auth flow and session

- Next auth receives a JWT token containing an identifier "sub" that uniquely
  identifies the user
- We create a user on our side, and save this sub so that we can find this
  user for subsequent logins
- When a chart is created, we check the session for the current user and
  add the user id if it exists

### Deployment

Environment variables to set when deploying :

- Set KEYCLOAK_ISSUER
- Set KEYCLOAK_SECRET
- Set NEXTAUTH_SECRET

- [ ] test.visualize.admin.ch
- [ ] int.visualize.admin.ch
- [x] vercel preview "auth"
